NameLayerMain/src/vg/civcraft/mc/namelayer/group/Group.java

```diff
@@ -317,23 +318,38 @@ public PlayerType getCurrentRank(UUID uuid) {
 	 * @param type- The PlayerType to add. If a preexisting PlayerType is found, 
 	 * it will be overwritten.
 	 */
-	public void addMember(UUID uuid, PlayerType type) {
+	
+	public void addMember(UUID uuid, PlayerType type){
+		addMember(uuid,type,true);
+	}
+	
+	public void addMember(UUID uuid, PlayerType type, boolean savetodb) {
 		if (type == PlayerType.NOT_BLACKLISTED) {
 			return;
 		}
-		if (isMember(uuid, type)) {
-			db.removeMember(uuid, name);
+		if (savetodb == true) {
+			if (isMember(uuid, type)){
+				db.removeMemberAsync(uuid, name);
```

Comment:
Bukkit uses an asynchronous threadpool of about 10+ worker threads; _generally_ yes your assumption might seem safe but _actually_ it's a liability as you can not be sure that a thread which _gets_ your task off the queue also _executes_ prior to the thread that gets the second task. 

Example: 
* Thread 1 polls queue, gets db.remove.
* Thread 1 is put on wait before running db.remove by the CPU scheduler
* Thread 2 polls queue, gets db.add.
* Thread 2 is immediately scheduled by the CPU for core 2, and executes.
* Thread 3 does some more work.
* Thread 1 is awoke by the CPU, and enters sync lock wait on DB connection handler b/c of Thread 2's lock.
* Thread 2 completes, but fails b/c multiple insert. A stack is printed.
* Thread 1 gets the lock on DB Connection handler and executes, the remove is succesful.
* The user is now gone from the group, instead of getting promoted.

Async Scheduling is not a kind mistress. If ordering matters, make it explicit, or refactor so ordering doesn't matter.